### PR TITLE
feat(resource): respecting CUDA_VISIBLE_DEVICES

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: aarnphm-ec2-dev <29749331+aarnphm@users.noreply.github.com>

## What does this PR address?

Fixes #(issue)

## Before submitting:

- [ ] Does the Pull Request follow [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary) naming? Here are [GitHub's
      guide](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) on how to create a pull request.
- [ ] Does the code follow BentoML's code style, `pre-commit run -a` script has passed ([instructions](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#style-check-auto-formatting-type-checking))?
- [ ] Did you read through [contribution guidelines](https://github.com/bentoml/BentoML/blob/main/CONTRIBUTING.md#ways-to-contribute) and follow [development guidelines](https://github.com/bentoml/BentoML/blob/main/DEVELOPMENT.md#start-developing)?
- [ ] Did your changes require updates to the documentation? Have you updated
      those accordingly? Here are [documentation guidelines](https://github.com/bentoml/BentoML/tree/main/docs) and [tips on writting docs](https://github.com/bentoml/BentoML/tree/main/docs#writing-documentation).
- [ ] Did you write tests to cover your changes?

### CODE REVIEW

File: src/bentoml/_internal/resource.py [hunk 1]
```diff
@@ -244,6 +213,13 @@ def from_system(cls) -> list[int]:
         """
         import pynvml  # type: ignore
 
+        cuda_visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES", None)
+        if cuda_visible_devices is not None:
+            if "," in cuda_visible_devices:
+                return [int(i) for i in cuda_visible_devices.split(",")]
+            else:
```
Comment (sauyon) [inline src/bentoml/_internal/resource.py:220]: Seems like the spec says comma separated (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars); why are we supporting space?

File: tests/unit/_internal/test_resource.py [hunk 2]
```diff
@@ -63,6 +63,14 @@ def test_CpuResource():
         CpuResource.from_spec((1, 2, 3))
 
 
+@pytest.mark.parametrize("env", ["0, 1", "0,1", "0 1"])
```
Comment (sauyon) [inline tests/unit/_internal/test_resource.py:66]: Maybe add a test for an empty value?

**Review by sauyon — COMMENTED** [review]
Seems like you haven't implemented `-1` (see https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#env-vars), we should support this properly.

Also, int casting is probably incorrect, we should allow strings, as per the spec.

File: tests/unit/_internal/test_resource.py (hunk 2 above)
Comment (aarnphm) [reply tests/unit/_internal/test_resource.py:66]: sg

File: src/bentoml/_internal/runner/strategy.py [hunk 3]
```diff
@@ -150,7 +151,12 @@ def get_worker_env(
                 ]
                 dev = ",".join(map(str, assigned_gpu))
             else:
-                dev = str(nvidia_gpus[worker_index // workers_per_resource])
+                idx = worker_index // workers_per_resource
+                if len(nvidia_gpus) == idx:
```
Comment (frostming) [inline src/bentoml/_internal/runner/strategy.py:155]:
```suggestion
                if idx >= len(nvidia_gpus):
```

File: src/bentoml/_internal/resource.py [hunk 4]
```diff
@@ -183,120 +188,92 @@ def query_os_cpu_count() -> int:
...
+                if spec == -1:
+                    return []
                 if spec < -1:
                     raise ValueError
-                return list(range(spec))
+                return [str(i) for i in range(spec)]
             elif isinstance(spec, str):
-                return cls.from_spec(int(spec))
+                try:
+                    return cls.from_spec(int(spec))
```
Comment (frostming) [inline src/bentoml/_internal/resource.py:208]: Better to convert str before evaluation to avoid recursive call.

Comment (aarnphm) [reply src/bentoml/_internal/resource.py:208]: Now thinking a bit more, I don't think it wouldn't recurse unless the spec is a valid int right?

Comment (aarnphm) [reply src/bentoml/_internal/resource.py:208]: Actually this would fail with `0,1,2`

File: src/bentoml/_internal/resource.py [hunk 5]
```diff
@@ -183,120 +188,95 @@ def query_os_cpu_count() -> int:
...
-    def from_spec(cls, spec: int | str | list[int | str]) -> list[int]:
+    def from_spec(cls, spec: int | str | list[int | str]) -> list[str]:
         if not isinstance(spec, (int, str, list)):
             raise TypeError(
                 "NVidia GPU device IDs must be int, str or a list specifing the exact GPUs to use."
             )
 
         try:
             if isinstance(spec, int):
+                if spec == -1:
```
Comment (sauyon) [inline src/bentoml/_internal/resource.py:201]: Actually, does `pynvml` not have support for parsing `CUDA_VISIBLE_DEVICES`? Maybe we can just use that?

File: src/bentoml/_internal/resource.py [hunk 6]
```diff
@@ -183,120 +188,95 @@ def query_os_cpu_count() -> int:
...
                     raise ValueError
-                return list(range(spec))
+                return [str(i) for i in range(spec)]
             elif isinstance(spec, str):
-                return cls.from_spec(int(spec))
+                try:
+                    resolved = int(spec)
+                    return cls.from_spec(resolved)
+                except ValueError:
+                    if spec.startswith("GPU"):
```
Comment (sauyon) [inline src/bentoml/_internal/resource.py:211]: I believe we should also support `MIG` strings. Let's maybe limit parsing to as little as possible so we don't have to keep up with changes in the Nvidia libraries.

File: src/bentoml/_internal/resource.py [hunk 7]
```diff
@@ -183,120 +188,95 @@ def query_os_cpu_count() -> int:
...
+                        return [spec]
+                    if "," in spec:
+                        return [i.strip() for i in spec.split(",")]
+                    raise ValueError
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply src/bentoml/_internal/resource.py:215]: We are catching it below right? I guess that was what we are doing b4. I can raise more specific error for this then.

[the remark this answers is not in the document]
Comment (sauyon) [reply src/bentoml/_internal/resource.py:215]: Oh, no need. Can just reraise in this case, I think it should make full traces look better anyway.

[the remark this answers is not in the document]
Comment (sauyon) [reply src/bentoml/_internal/resource.py:215]:
```suggestion
                    raise
```

File: src/bentoml/_internal/resource.py
Comment (bojiang) [inline src/bentoml/_internal/resource.py:1]: @aarnphm Shall we add some unit tests for this CUDA_VISIBLE_DEVICES resolver?
I guess you already know this doc,
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-environment-variables
maybe we should add test cases in this doc

**Comment by GrigoryEvko** [comment]
Guys how this wasn't implemented? How do I run exactly one service per one gpu? bentoml doesn't respect env variables, I can't configure it inside bentoml service? so how??
